### DISCUSSION TITLE
Service task output

### DISCUSSION THREAD

**Discussion by linonetwo** [bpmnServer/bpmn-server#218] [Q&A] [answered]
Hi, do you know how to pipe one service task's output as next's input? Just like AI workflow like defy, coze.

Is previous service task's output variable available in next task's input expression? And how to gather input from multiple "previous node"?

**Comment by ralphhanna** [comment]
I am not familiar with AI workflow, can you please explain?

Currently this is done through scripting and use of instance `data` please see https://bpmnserver.github.io/docs/data/#inputoutput-using-scripts

**Reply by linonetwo** [reply]
For example in this bpmn

![图片](https://github.com/user-attachments/assets/b8ec86b6-16d9-4dd7-ba32-5b8092b3ee44)

```ts
import { Execution, IServiceProvider } from 'bpmn-server';

export const twServiceProvider: IServiceProvider = {
  'log-console'(input: Record<string, unknown>, _execution: Execution) {
    console.log('log-console service called:', input);
    return {
      a: 'b',
    }
  },
  a: {
    b: (input: Record<string, unknown>, execution: Execution) => {
      console.log(`b is called with ${JSON.stringify(input)} and last output is ${JSON.stringify(execution.item.output)}`);
    },
  },
};

```

first service task called the `log-console` and it returns `{ a: 'b'}`, I want to get it in second service task called `a.b`.

I find the return value is placed in `item.output`, but I can't get previous item's output. (I think item means service task node?)

https://github.com/bpmnServer/bpmn-server/blob/682d4cffa9f4ea908aac4bf5169956c4501cc9d4/src/elements/Tasks.ts#L84-L90

I want to achieve data-flow https://docs.camunda.io/docs/components/modeler/bpmn/data-flow/ , where second worker can get result data from previous workers (one or many).

About AI workflow, one service task might call OpenAi API, and get LLM result test, it put the text in result. And second service task will call OpenAi API with previous result text again to shorten the result. This kind of chainned AI is called Agent these days.

**Comment by linonetwo** [comment]
This works `execution.item.token.lastItem.output`

```ts
export const twServiceProvider: IServiceProvider = {
  'log-console'(input: Record<string, unknown>, _execution: Execution) {
    console.log('log-console service called:', input);
    return {
      a: 'b',
    }
  },
  a: {
    b: (input: Record<string, unknown>, execution: Execution) => {
      console.log(`b is called with ${JSON.stringify(input)} and last output is ${JSON.stringify(execution.item.token.lastItem.output)}`);
    },
  },
};
```

log

```
log-console service called: { str: 'The text' }
b is called with {} and last output is {"a":"b"}
```

**Comment by linonetwo** [comment]
Another question is what does the output on the panel mean? I didn't see where it is used in the `ScriptTask.run`

![图片](https://github.com/user-attachments/assets/00a676c1-5eda-48ad-8166-4c297272f9f4)

So does the input in the panel.

And exclusive gateway's expression, where does the variable comes from?

https://docs.camunda.io/docs/components/concepts/workflow-patterns/#conditions-ifthen

I simply want to create something like defi and coze.

**Comment by ralphhanna** [comment]
I really like to know what does defi and coze do??

in bpmn-server, output variable is what is returned from the activity into item.output

**Reply by linonetwo** [reply]
I think they works exactly like noflo, see https://github.com/noflo or  https://app.flowhub.io/ . You can define node to send API request or do DB search, and use a node to merge the result, and pass to AI API to generate summary, and return to user.

Sadly noflo is dead, doesn't live as long as BPMN.



> returned from the activity into item.output

I'm sure about this, it is what I return in the code. But not what I defined on the properties panel. So the output definition on the properties panel is truely not being used in bpmn-server?

**Comment by ralphhanna** [comment]
There are two problems with the above solution **execution.item.token.lastItem.output**
1. the lastItem may not be in the same token 
2. input and output variables are not persistent, so if there is a wait state the value of output is not preserved

So a more sure solution is using `instance.data` to hold transient data

**Reply by linonetwo** [reply]
I see, camunda works in this way, it put everything on a global context

![图片](https://github.com/user-attachments/assets/2fd07055-abc1-4da9-b25f-daebbe1afb30)

**Comment by ralphhanna** [comment, marked as answer]
In addition to above, Release 2.2.10, stores `item.output` so you can search back for the item and get its output.

Thanks
